File: packages/capacitor-notificare-in-app-messaging/android/src/main/java/re/notifica/iam/capacitor/NotificareInAppMessagingPlugin.kt
package re.notifica.iam.capacitor

import android.os.Handler
import android.os.Looper
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import re.notifica.Notificare
import re.notifica.iam.NotificareInAppMessaging
import re.notifica.iam.ktx.inAppMessaging
import re.notifica.iam.models.NotificareInAppMessage
import re.notifica.internal.NotificareLogger

@CapacitorPlugin(name = "NotificareInAppMessagingPlugin")
public class NotificareInAppMessagingPlugin : Plugin(), NotificareInAppMessaging.MessageLifecycleListener {

    override fun load() {
        EventBroker.setup(this::notifyListeners)

        Notificare.inAppMessaging().removeLifecycleListener(this)
        Notificare.inAppMessaging().addLifecycleListener(this)
    }

    // region Notificare In-App Messaging

    @PluginMethod
    public fun hasMessagesSuppressed(call: PluginCall) {
        try {
            call.resolve(
                JSObject().apply {
                    put("result", Notificare.inAppMessaging().hasMessagesSuppressed)
                }
            )
        } catch (e: Exception) {
            call.reject(e.localizedMessage)
        }
    }

    @PluginMethod
    public fun setMessagesSuppressed(call: PluginCall) {
        try {
            val suppressed = call.getBoolean("suppressed") ?: run {
                call.reject("Missing 'product' parameter.")
                return
            }

            Notificare.inAppMessaging().hasMessagesSuppressed = suppressed
            call.resolve()
        } catch (e: Exception) {
            call.reject(e.localizedMessage)
        }
    }

    // region NotificareInAppMessaging.MessageLifecycleListener

    override fun onMessagePresented(message: NotificareInAppMessage) {
        try {
            val data = JSObject().apply {
                put("message", message.toJson())
            }

            EventBroker.dispatchEvent("message_presented", data)
        } catch (e: Exception) {
            NotificareLogger.error("Failed to emit the message_presented event.", e)
        }
    }

    override fun onMessageFinishedPresenting(message: NotificareInAppMessage) {
        try {
            val data = JSObject().apply {
                put("message", message.toJson())
            }

            EventBroker.dispatchEvent("message_finished_presenting", data)
        } catch (e: Exception) {
            NotificareLogger.error("Failed to emit the message_finished_presenting event.", e)
        }
    }

    override fun onMessageFailedToPresent(message: NotificareInAppMessage) {
        try {
            val data = JSObject().apply {
                put("message", message.toJson())
            }

            EventBroker.dispatchEvent("message_failed_to_present", data)
        } catch (e: Exception) {
            NotificareLogger.error("Failed to emit the message_failed_to_present event.", e)
        }
    }

    override fun onActionExecuted(message: NotificareInAppMessage, action: NotificareInAppMessage.Action) {
        try {
            val data = JSObject().apply {
                put("message", message.toJson())
                put("action", action.toJson())
            }

            EventBroker.dispatchEvent("action_executed", data)
        } catch (e: Exception) {
            NotificareLogger.error("Failed to emit the action_executed event.", e)
        }
    }

    override fun onActionFailedToExecute(
        message: NotificareInAppMessage,
        action: NotificareInAppMessage.Action,
        error: Exception?
    ) {
        try {
            val data = JSObject().apply {
                put("message", message.toJson())
                put("action", action.toJson())
            }

            if (error != null) {
                data.put("error", error.message)
            }

            EventBroker.dispatchEvent("action_failed_to_execute", data)
        } catch (e: Exception) {
            NotificareLogger.error("Failed to emit the action_failed_to_execute event.", e)
        }
    }

    // endregion

    // endregion

    public companion object {
        internal fun onMainThread(action: () -> Unit) = Handler(Looper.getMainLooper()).post(action)
    }
}

File: packages/capacitor-notificare-in-app-messaging/ios/Plugin/NotificareInAppMessagingPlugin.swift
import Foundation
import Capacitor
import NotificareKit
import NotificareInAppMessagingKit

@objc(NotificareInAppMessagingPlugin)
public class NotificareInAppMessagingPlugin: CAPPlugin {
    public override func load() {
        EventBroker.instance.setup { self.notifyListeners($0, data: $1) }
        Notificare.shared.inAppMessaging().delegate = self
    }
    
    @objc func hasMessagesSuppressed(_ call: CAPPluginCall) {
        call.resolve([
            "result": Notificare.shared.inAppMessaging().hasMessagesSuppressed
        ])
    }
    
    @objc func setMessagesSuppressed(_ call: CAPPluginCall) {
        guard let suppressed = call.getBool("suppressed") else {
            call.reject("Missing 'suppressed' parameter.")
            return
        }
        
        let evaluateContext = call.getBool("evaluateContext") ?? false
        
        Notificare.shared.inAppMessaging().setMessagesSuppressed(suppressed, evaluateContext: evaluateContext)
        
        call.resolve()
    }
}

extension NotificareInAppMessagingPlugin: NotificareInAppMessagingDelegate {
    public func notificare(_ notificare: NotificareInAppMessaging, didPresentMessage message: NotificareInAppMessage) {
        do {
            let data = [
                "message": try message.toJson()
            ]

            EventBroker.instance.dispatchEvent("message_presented", data: data)
        } catch {
            NotificareLogger.error("Failed to emit the message_presented event.", error: error)
        }
    }

    public func notificare(_ notificare: NotificareInAppMessaging, didFinishPresentingMessage message: NotificareInAppMessage) {
        do {
            let data = [
                "message": try message.toJson()
            ]

            EventBroker.instance.dispatchEvent("message_finished_presenting", data: data)
        } catch {
            NotificareLogger.error("Failed to emit the message_finished_presenting event.", error: error)
        }
    }

    public func notificare(_ notificare: NotificareInAppMessaging, didFailToPresentMessage message: NotificareInAppMessage) {
        do {
            let data = [
                "message": try message.toJson()
            ]

            EventBroker.instance.dispatchEvent("message_failed_to_present", data: data)
        } catch {
            NotificareLogger.error("Failed to emit the message_failed_to_present event.", error: error)
        }
    }

    public func notificare(_ notificare: NotificareInAppMessaging, didExecuteAction action: NotificareInAppMessage.Action, for message: NotificareInAppMessage) {
        do {
            let data = [
                "message": try message.toJson(),
                "action": try action.toJson(),
            ]

            EventBroker.instance.dispatchEvent("action_executed", data: data)
        } catch {
            NotificareLogger.error("Failed to emit the action_executed event.", error: error)
        }
    }

    public func notificare(_ notificare: NotificareInAppMessaging, didFailToExecuteAction action: NotificareInAppMessage.Action, for message: NotificareInAppMessage, error: Error?) {
        do {
            var data: [String: Any] = [
                "message": try message.toJson(),
                "action": try action.toJson(),
            ]

            if let error = error {
                data["error"] = error.localizedDescription
            }

            EventBroker.instance.dispatchEvent("action_failed_to_execute", data: data)
        } catch {
            NotificareLogger.error("Failed to emit the action_failed_to_execute event.", error: error)
        }
    }
}
